File: src/gui/framerate.cpp
//
// Created by Mike on 2021/11/30.
//

#include <gui/framerate.h>

namespace luisa::compute {

Framerate::Framerate(size_t n) noexcept
    : _history_size{n} {
    _durations.reserve(n);
    _frames.reserve(n);
    _last = Clock::now();
}

void Framerate::clear() noexcept {
    _durations.clear();
    _frames.clear();
    _last = Clock::now();
}

double Framerate::duration() const noexcept {
    auto dt = Clock::now() - _last;
    using namespace std::chrono_literals;
    return static_cast<double>(dt / 1ns) * 1e-9;
}

void Framerate::record(size_t frame_count) noexcept {
    if (_durations.size() == _history_size) {
        _durations.erase(_durations.begin());
        _frames.erase(_frames.begin());
    }
    using namespace std::chrono_literals;
    _durations.emplace_back(duration());
    _frames.emplace_back(frame_count);
    _last = Clock::now();
}

double Framerate::report() const noexcept {
    if (_durations.empty()) { return 0.0; }
    auto total_duration = 0.0;
    auto total_frame_count = static_cast<size_t>(0u);
    for (auto i = 0u; i < _durations.size(); i++) {
        total_duration += _durations[i];
        total_frame_count += _frames[i];
    }
    return static_cast<double>(total_frame_count) / total_duration;
}

}

File: src/py/luisa/framerate.py
import time
from functools import reduce

class FrameRate:
    def __init__(self, history_size: int) -> None:
        self.history_size: int = history_size
        self._durations_and_frames: list[tuple[float, int]] = []
        self.last: float = time.time()
    
    def clear(self) -> None:
        self._durations_and_frames.clear()
        self.last = time.time()

    def duration(self) -> float:
        return time.time() - self.last

    def record(self, frame_count: int = 1) -> None:
        if len(self._durations_and_frames) == self.history_size:
            self._durations_and_frames.pop(0)
        self._durations_and_frames.append((self.duration(), frame_count))
        self.last = time.time()

    def report(self) -> float:
        if len(self._durations_and_frames) == 0:
            return 0
        tuple_add = lambda x, y: (x[0] + y[0], x[1] + y[1])
        total_duration, total_frame_count = reduce(tuple_add, self._durations_and_frames)
        return total_frame_count / total_duration
